EpiAware/src/epimodel.jl

```diff
@@ -1,96 +1,83 @@
-
-"""
-struct EpiModel{T<:Real} <: AbstractEpiModel
+struct EpiData{T<:Real,F<:Function}
     gen_int::Vector{T}
     delay_int::Vector{T}
     delay_kernel::SparseMatrixCSC{T,Integer}
     cluster_coeff::T
-    len_gen_int::Integer #length(gen_int) just to save recalc
-    len_delay_int::Integer #length(delay_int) just to save recalc
+    len_gen_int::Integer
+    len_delay_int::Integer
     time_horizon::Integer
+    transformation::F
 
-    #Inner constructors for EpiModel object
-    function EpiModel(gen_int, delay_int, cluster_coeff, time_horizon::Integer)
+    #Inner constructors for EpiData object
+    function EpiData(
+        gen_int,
+        delay_int,
+        cluster_coeff,
+        time_horizon::Integer,
+        transformation::Function,
+    )
         @assert all(gen_int .>= 0) "Generation interval must be non-negative"
         @assert all(delay_int .>= 0) "Delay interval must be non-negative"
         @assert sum(gen_int) ≈ 1 "Generation interval must sum to 1"
         @assert sum(delay_int) ≈ 1 "Delay interval must sum to 1"
 
         K = generate_observation_kernel(delay_int, time_horizon)
 
-        new{eltype(gen_int)}(
+        new{eltype(gen_int),typeof(transformation)}(
             gen_int,
             delay_int,
             K,
             cluster_coeff,
             length(gen_int),
             length(delay_int),
             time_horizon,
+            transformation,
         )
     end
 
-    function EpiModel(
+    function EpiData(
         gen_distribution::ContinuousDistribution,
         delay_distribution::ContinuousDistribution,
         cluster_coeff,
         time_horizon::Integer;
-        Δd = 1.0,
         D_gen,
         D_delay,
+        Δd = 1.0,
+        transformation::Function = exp,
     )
         gen_int =
             create_discrete_pmf(gen_distribution, Δd = Δd, D = D_gen) |>
             p -> p[2:end] ./ sum(p[2:end])
         delay_int = create_discrete_pmf(delay_distribution, Δd = Δd, D = D_delay)
 
-        K = generate_observation_kernel(delay_int, time_horizon)
-
-        new{eltype(gen_int)}(
-            gen_int,
-            delay_int,
-            K,
-            cluster_coeff,
-            length(gen_int),
-            length(delay_int),
-            time_horizon,
-        )
+        return EpiData(gen_int, delay_int, cluster_coeff, time_horizon, transformation)
     end
 end
 
-"""
-    (epi_model::EpiModel)(recent_incidence, Rt)
+struct DirectInfections <: AbstractEpiModel
+    data::EpiData
+end
 
-Apply the EpiModel to calculate new incidence based on recent incidence and Rt.
+function (epi_model::DirectInfections)(recent_incidence, _I_t)
+    nothing, epi_model.data.transformation(_I_t)
+end
+
+struct ExpGrowthRate <: AbstractEpiModel
+    data::EpiData
+end
 
-# Arguments
-- `recent_incidence`: Array of recent incidence values.
-- `Rt`: Reproduction number.
+function (epi_model::ExpGrowthRate)(recent_incidence, rt)
+    new_incidence = recent_incidence * exp(rt)
```

Comment:
does it make any difference in julia to do this as exp(log_recent_incidence + rt) vs current suggestion?

Reply:
Also this implementation suggests it will be used interatively (in the scan function). Are we clear that this is more efficient than the usual approach to implementation (exp(log_init_incidence + cumsum(rt)))?

Reply:
No. That would be fine.

Reply:
In fact, I just noted that I should have been treating the initial value generated by `random_walk` and the initialiser for the infection time series. 

So above was wrong. Just committed the fix.

Reply:
For both?

Reply:
> For both?

Hmmmm...... Actually we think abit about the `init` for renewal. As ever.